File: algorithms/2_medium/F/00640_solve_the_equation/solution.c
typedef ptrdiff_t isize;

enum extract_state {
    IS_NUM,
    IS_X,
    IS_EQ,
    NONE
};

struct extract_result {
    enum extract_state state;
    int32_t val;
    isize new_start_idx;
};

static inline struct extract_result extract_nums(const char* s, isize idx)
{
    if (s[idx] == '\0') {
        return (struct extract_result){ NONE, 0, 0 };
    }

    if (s[idx] == '=') {
        return (struct extract_result){ IS_EQ, 0, idx + 1 };
    }

    bool is_pos = true;
    if (s[idx] == '+') {
        idx += 1;
    } else if (s[idx] == '-') {
        is_pos = false;
        idx += 1;
    } else if ((s[idx] > '9' || s[idx] < '0') && s[idx] != 'x') {
        fprintf(stderr, "invalid input");
        exit(1);
    }

    if (s[idx] == 'x') {
        return (struct extract_result){ IS_X, (is_pos ? 1 : -1), idx + 1 };
    }

    int32_t num = 0;
    while (s[idx] >= '0' && s[idx] <= '9') {
        num = num * 10 + s[idx] - '0';
        idx += 1;
    }

    if (s[idx] == 'x') {
        return (struct extract_result){ IS_X, (is_pos ? num : -num), idx + 1 };
    }

    return (struct extract_result){ IS_NUM, (is_pos ? num : -num), idx };
}

char* solveEquation(char* equation)
{
    isize start_idx = 0;
    bool is_left_op = true;
    int32_t num = 0;
    int32_t x = 0;
    bool no_more_tokens = false;
    while (!no_more_tokens) {
        struct extract_result res = extract_nums(equation, start_idx);

        switch (res.state) {
            case (IS_NUM):
                num += is_left_op ? res.val : -res.val;
                start_idx = res.new_start_idx;
                break;
            case (IS_X):
                x += is_left_op ? res.val : -res.val;
                start_idx = res.new_start_idx;
                break;
            case IS_EQ:
                is_left_op = false;
                start_idx = res.new_start_idx;
                break;
            case NONE:
                no_more_tokens = true;
                break;
            default:
                break;
        }
    }

    char* res = malloc(30 * sizeof(*res));
    assert(res != NULL);

    if (x == 0) {
        if (num == 0) {
            snprintf(res, 30, "Infinite solutions");
        } else {
            snprintf(res, 30, "No solution");
        }
    } else {
        snprintf(res, 30, "x=%d", -num / x);
    }

    return res;
}

File: algorithms/2_medium/F/00640_solve_the_equation/solution.rs
enum ExtractResult {
    Num(i32, usize),
    X(i32, usize),
    EqualSign(usize),
    None,
}

impl Solution {
    #[inline(always)]
    fn extract_nums(bytes: &[u8], mut idx: usize) -> ExtractResult {
        if idx >= bytes.len() {
            return ExtractResult::None;
        }

        if bytes[idx] == b'=' {
            return ExtractResult::EqualSign(idx + 1);
        }

        let is_pos = match bytes[idx] {
            b'+' => {
                idx += 1;
                true
            },
            b'-' => {
                idx += 1; 
                false
            },
            b'0'..=b'9' | b'x' => true,
            _ => unreachable!(),
        };

        if bytes[idx] == b'x' {
            return ExtractResult::X(if is_pos { 1 } else { -1 }, idx + 1);
        }

        let mut num = 0;
        while let Some(&byte) = bytes.get(idx) && byte >= b'0' && byte <= b'9' {
            num = num * 10 + (byte - b'0') as i32;
            idx += 1;
        }

        match bytes.get(idx) {
            Some(b'x') => ExtractResult::X(if is_pos { num } else { -num }, idx + 1),
            _ => ExtractResult::Num(if is_pos { num } else { -num }, idx)
        }
    }

    pub fn solve_equation(equation: String) -> String {
        let bytes = equation.as_bytes();
        let mut start_idx = 0_usize;
        let (mut num, mut x, mut is_left_op) = (0, 0, true);
        while let state = Self::extract_nums(bytes, start_idx) {
            match state {
                ExtractResult::Num(val, new_start_idx) => {
                    num += if is_left_op { val } else { -val };
                    start_idx = new_start_idx;
                },
                ExtractResult::X(val, new_start_idx) => {
                    x += if is_left_op { val } else { -val };
                    start_idx = new_start_idx;
                },
                ExtractResult::EqualSign(new_start_idx) => {
                    is_left_op = false;
                    start_idx = new_start_idx;
                },
                ExtractResult::None => {
                    break;
                }
            }
        }

        if x == 0 {
            return String::from(if num == 0 { "Infinite solutions" } else { "No solution" });
        }

        num /= x;
        format!("x={}", -num)
    }
}
